Document metrics through OpenTelemetry (#65)
kubewarden/docs: src/operator-manual/metrics/01-quickstart.md

# Metrics

This section describes how to enable metrics reporting on the Policy Server.

We are going to use [Prometheus](https://prometheus.io/) to scrape metrics exposed by the Policy
Server.

> ## Install Prometheus

Reviewer: Maybe we should define a new 2nd level section called `Prerequisites` with the following 3rd level sections:

  1. Install OpenTelemetry
  2. Install Prometheus
  3. Install Kubewarden

Author: This is already the case, if I'm not misunderstanding your comment.

The index looks like:

7. Operator Manual
  7.1. OpenTelemetry: intructions on setting up a Kubernetes cluster + opentelemetry
  7.2.  Metrics: setting up jaeger + kubewarden with jaeger integration
  7.3. Tracing: setting up prometheus + kubewarden with prometheus integration

Or do you refer to something else?

Reviewer 2: Oh true, the numbering only appears on the rendered book. I'm happy with that.

Reviewer: I see. I would still add some text at the top of this file to ensure people are aware of the requirements that have been covered by the other sections of the book.

To put it in a more direct way: ensure you went through "OpenTelemetry" section before doing anything written here

We will use the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator),
that allows us to intuitively define Prometheus' Targets.

> There are many ways to install and set up prometheus. For ease of deployment, we will use the

Reviewer: ```suggestion
There are many ways to install and set up Prometheus. For ease of deployment, we will use the
```

Prometheus community helm chart.

Let's add the helm repository from the Prometheus Community:

```console
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
```

Now, let's install the
[`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack)
chart. This chart contains a collection of Kubernetes manifests, Grafana dashboards, and Prometheus
rules.

Let's create a `kube-prometheus-stack-values.yaml` file to configure the `kube-prometheus-stack`
Helm chart values with the following contents:

```yaml
prometheus:
  additionalServiceMonitors:
    - name: kubewarden
      selector:
        matchLabels:

>           app: kubewarden-policy-server-default

Reviewer: Unrelated with this PR, but still... Shouldn't we use a more generic name here? I'm thinking about when we will have the Kubewarden controller expose metrics too. I guess it will have to use the same label in order to be scraped by Prometheus.

Author: This label is added by the controller, we are just targeting it here. https://github.com/kubewarden/kubewarden-controller/blob/9d0d1a02817eb12f82fffd5ea4a792b15cd35baf/internal/pkg/admission/policy-server-service.go#L30.

      namespaceSelector:
        matchNames:
          - kubewarden
      endpoints:
        - port: metrics
          interval: 10s
```

The `prometheus-operator` deployed as part of this Helm chart defines the concept of [Service
Monitors](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/design.md#servicemonitor),

> used to declaratively define what services should be monitored by Prometheus.

Reviewer: ```suggestion
used to declaratively define which services should be monitored by Prometheus.
```

In our case, we are adding a Service monitor targeting the `kubewarden` namespace, for services that

> match labels `app=kubewarden-policy-server-ha-policy-server`. This way, the Prometheus Operator is

Reviewer: The label provided here is different from the one specified in the yaml example from above

> able to inspect what `endpoints` are tied to services matching this conditions. The operator will

Reviewer: ```suggestion
able to inspect which Kubernetes Endpoints are tied to services matching this conditions. The operator will
```

> then take care of generating a valid configuration file for Prometheus, and reload Prometheus

Reviewer: ```suggestion
then take care of generating a valid configuration file for Prometheus, and reloading it
```

```console
helm install --wait --create-namespace --namespace prometheus --values kube-prometheus-stack-values.yaml prometheus prometheus-community/kube-prometheus-stack
```

## Install Kubewarden

We can now install Kubewarden in the recommended way with the Helm chart.

> **Note well:** cert-manager and OpenTelemetry are a requirement of Kubewarden, but we've already
> installed them in a previous section of this book.

As a first step, we have to add the Helm repository that contains Kubewarden:

```console
helm repo add kubewarden https://charts.kubewarden.io
```

Then we have to install the Custom Resource Definitions (CRDs) defined by
Kubewarden:

```console
helm install kubewarden-crds kubewarden/kubewarden-crds
```

Now we can deploy the rest of the Kubewarden stack. The official helm chart will create a
PolicyServer named `default`. The default Helm chart values are already set up in a way that metrics
are exposed for this default Policy Server. The relevant Helm Chart values are:

```yaml
policyServer
  telemetry:
    metrics:
      port: 8080
```

Note that the default `values.yaml` for this Helm Chart are already set to this values. There is no

> need set this values if you don't want to change anything else.

Reviewer: Just out of curiosity, is everything fine if the user doesn't have OpenTelemetry installed?

Author: Just tried out of curiosity.

If OpenTelemetry is not installed and on the `values.yaml` they provide:

```yaml
policyServer:
  telemetry:
    enabled: False
```

Everything is as expected. No unwanted behaviors anywhere I could see.

If OpenTelemetry is left enabled (default `values.yaml`), but OpenTelemetry is not installed, we get an installation error with Helm because the `OpenTelemetryCollector` resource kind is not registered in the API server. I think this is fine and desirable.

Reviewer 2: should we default to OpenTelemetry disabled on the chart, for an easier first-use? (if not, we would need to change docs of charts, plus the docs page, for all deployments of kubewarden).

Author: I think we should default to telemetry enabled by default with Helm. Any kind of deployment on top of Kubernetes requires some visibility.

I am actually missing on this PR to update the quickstart where the user disables it. Adding that.

Reviewer 2: Fair enough. Commented on the chart PR, that the chart upgrades would need a release notice.

Author: Updated the PR, thanks :)

We can install the helm chart:

```console
helm install --wait --create-namespace --namespace kubewarden kubewarden-controller kubewarden/kubewarden-controller
```

This leads to the creation of the `default` instance of `PolicyServer`:

```console
kubectl get policyservers.policies.kubewarden.io
NAME      AGE
default   3m7s
```

By default, this policy server will have metrics enabled.

> ## Enforcing a policy

Reviewer: I have some doubts about this section. I would skip it tbh

Reviewer 2: I agree.

Author: Should we skip it on tracing too? I wanted to show a way to make the counters go up and be able to inspect how they grow.

Reviewer 2: Would it be enough just explaining in a paragraph what is expected to happen?
To be fair, I have no problem with it as an example after the whole set up, but the maintenance of the examples without the single source of truth may get annoying.

Reviewer: I agree :100:% with what @viccuad said.

EOF
```

We can wait for the policy to be active in this way:

```console
kubectl wait --for=condition=PolicyActive clusteradmissionpolicy/mandatory-owner
```

Once the policy is active, we can try it out in this way:

```console
kubectl apply -f - <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: team-beta-prod
  labels:
    owner: some-owner
spec:
  selector:
    matchLabels:
      app: nginx
  replicas: 0
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
EOF
```

This Deployment object will be created because it doesn't violate the policy.

On the other hand, this Deployment will be blocked by the policy:

```console
kubectl apply -f - <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment-without-labels
  namespace: team-beta-prod
spec:
  selector:
    matchLabels:
      app: nginx
  replicas: 0
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
EOF
```

> One of the metrics that the `PolicyServer` exposes is the number or policy evaluations. We can check

Reviewer: I would create a new section like `Accessing Prometheus` with this information,
plus another section about `Accessing Grafana`.

They don't have to dig into the details, just provide the URLs (and eventual `kubectl port-forward` command) required to access them
